### PR TITLE
wasm-smith: allow generating single function modules

### PR DESCRIPTION
**Pull request by abrown**
This change adds `SingleFunctionModule`, an alternative to `Module` and
`ConfiguredModule` that allows the user to specify the function
signature to generate. It avoids previous issues with `Module` where
`min_funcs` and `max_funcs` would be configured as `1` but the function
would at times not be exported.

### CODE REVIEW

File: crates/wasm-smith/src/core.rs [hunk 1]
```diff
@@ -233,6 +233,68 @@ impl<'a> Arbitrary<'a> for MaybeInvalidModule {
...
+            .exports
+            .push(("generated_function".into(), ExportKind::Func, 0));
+        self.module.arbitrary_exports(u)?;
+
+        self.module.arbitrary_start(u)?;
+        self.module.arbitrary_elems(u)?;
+        self.module.arbitrary_data(u)?;
+
+        // No need to modify the generated code.
+        self.module.arbitrary_code(u, allow_invalid)?;
```
Comment (abrown) [inline crates/wasm-smith/src/core.rs:293]: This comment is probably incorrect: I think there is a need to modify the generated code to use the function parameters and eventually emit the desired results.

File: crates/wasm-smith/src/core.rs [hunk 2]
```diff
@@ -233,6 +233,68 @@ impl<'a> Arbitrary<'a> for MaybeInvalidModule {
     }
 }
 
+pub struct SingleFunctionModule {
+    signature: FuncType,
```
Comment (abrown) [inline crates/wasm-smith/src/core.rs:237]: This `FuncType` could be generated with `arbitrary_func_type`.

File: crates/wasm-smith/src/core.rs [hunk 3]
```diff
@@ -233,6 +233,68 @@ impl<'a> Arbitrary<'a> for MaybeInvalidModule {
...
+    pub fn build(&mut self, u: &mut Unstructured, allow_invalid: bool) -> Result<()> {
+        let ty = Type::Func(self.signature.into());
+        self.module.valtypes = configured_valtypes(&*self.module.config);
+
+        // Record the type of the single function. Since the module should only
+        // have one function, we do not import additional functions (i.e.,
+        // `arbitrary_imports()`) or other function types (i.e.,
+        // `arbitrary_types()`).
+        self.module.types.push(ty);
+        let ty_idx = self.module.types.len().try_into().unwrap();
```
Comment (abrown) [inline crates/wasm-smith/src/core.rs:263]: Obviously this kind of thing messes with the meaning of `Config::min_funcs` and `Config::max_funcs`. This `SingleFunctionModule` could grow its own `Config` object but it sure would be nice to figure out a way to re-use the existing `Config`.

**Comment by alexcrichton** [comment]
I like the idea of reusing `Module` where possible since it's got quite a lot internally already. If we can though I think I'd prefer to continue to lean heavily on `Config`. I'm not too too concerned about trying to rationalize all the configuration options with one another so if something falls through the cracks there that seems reasonable to me. 

I think though the main purpose of this would be to generate function signatures differently though, right? In that we'd generate arbitrary parameters but we wouldn't know the results until after the function was generated (presumably taking all of the results from the function)

**Comment by abrown** [comment]
> I think though the main purpose of this would be to generate function signatures differently though, right? In that we'd generate arbitrary parameters but we wouldn't know the results until after the function was generated (presumably taking all of the results from the function)

You mentioned this in the other thread but I didn't quite understand the "why." Is it because not restricting the result types would be easier to implement or something else? It seems valuable to provide a target for the function to "head towards" but perhaps that is too difficult (I haven't really looked at the code builder changes yet).

**Comment by fitzgen** [comment]
> but the function would at times not be exported.

We could also add a `export_all` method that performs a post-generation pass to export everything (or `export_all_funcs` and just export the functions)

**Comment by fitzgen** [comment]
> One of my worries about the current generation of wasm-smith is that we generate a bunch of code which has all these nice values on the wasm stack, only to have them all get dropped at the end of the function because the function has a different signature. That means that differential fuzzing is pretty unlikely to actually pick up any differences in execution since the values going in/out are always canned.
> 
> I don't know how best to fix it but picking the result types later was a strawman at doing so. It's still not great though I think.

Another thing we could do is, when we have a memory, write the extra values to memory instead of dropping them.

**Comment by abrown** [comment]
> That's sort of another thing which makes me think we should continue to add heuristics and config options but probably no need to make a whole new generator just yet.

Yeah, thinking about this more, a combination of [`export_everything`](https://github.com/bytecodealliance/wasm-tools/pull/657), [`allowed_instructions`](https://docs.rs/wasm-smith/0.11.1/wasm_smith/trait.Config.html#method.allowed_instructions), and whatever we come up with for https://github.com/bytecodealliance/wasm-tools/issues/655 is probably enough to generate small functions with observable results.

**Pull request closed without merging**